bottomless2/src/storage/mod.rs

```diff
@@ -0,0 +1,78 @@
+use chrono::{DateTime, Utc};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+use crate::NamespaceName;
+use crate::Result;
+
+mod fs;
+mod s3;
+
+pub struct SegmentMeta {
+    namespace: NamespaceName,
+    start_frame_no: u64,
+    end_frame_no: u64,
+    create_at: DateTime<Utc>,
+}
+
+pub struct RestoreRequest {}
+
+pub struct RestoreOptions {
+    /// Namespace to restore
+    namespace: NamespaceName,
+    /// If provided, will restore up to the most recent segment lesser or equal to `before`
+    before: Option<DateTime<Utc>>,
+}
+
+pub struct DbMeta {
+    max_frame_no: u64,
+}
+
+pub trait Storage: Send + Sync {
```

Comment:
I think we can use https://docs.rs/trait-variant/0.1.2/trait_variant/ to make the async fn sendable.

Reply:
we'll see it's that's necessary later